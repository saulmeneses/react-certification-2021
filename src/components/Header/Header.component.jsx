import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  Toolbar,
  IconButton,
  FormControlLabel,
  MenuItem,
  Drawer,
  ListItem,
  ListItemIcon,
  ListItemText,
} from '@material-ui/core';
import {
  Menu as MenuIcon,
  Search as SearchIcon,
  AccountCircle as AccountCircleIcon,
  Home as HomeIcon,
} from '@material-ui/icons';

import {
  StyledGrowDiv,
  StyledAppBar,
  StyledIconButton,
  StyledList,
  StyledSearchDiv,
  SearchIconDiv,
  StyledForm,
  StyledInputBase,
  StyledInput,
  StyledDarkModeDiv,
  StyledSwitch,
  StyledMenu,
} from './Header.styles';
import LoginDialog from '../LoginDialog';

import { useCustom } from '../../providers/Custom';
import { useYoutubeSearch } from '../../utils/hooks';

const Header = () => {
  const history = useHistory();
  const [openLoginDialog, setOpenLoginDialog] = useState(false);
  const [openDrawer, setOpenDrawer] = useState(false);
  const [anchorEl, setAnchorEl] = useState(null);
  const {
    darkMode,
    searchTerm,
    switchDarkMode,
    updateSearchTerm,
    updateSearchResult,
  } = useCustom();
  const yt = useYoutubeSearch({ type: 'video' });

  const isMenuOpen = Boolean(anchorEl);
  const handleGoBack = () => history.push('/');
  const handleMenuOpen = (e) => setAnchorEl(e.currentTarget);
  const handleMenuClose = () => setAnchorEl(null);
  const toggleDrawer = (open) => (e) => {
    if (e.type === 'keydown' && (e.key === 'Tab' || e.key === 'Shift')) {
      return;
    }

    setOpenDrawer(open);
  };
  const handleSearchOnChange = ({ target: { value } }) => updateSearchTerm(value);
  const handleSearchOnSubmit = (e) => {
    e.preventDefault();

    yt.search(searchTerm).then(updateSearchResult);
  };
  const handleDialogOpen = () => {
    handleMenuClose();
    setOpenLoginDialog(true);
  };
  const handleDialogClose = () => setOpenLoginDialog(false);

  return (
    <>
      <LoginDialog open={openLoginDialog} onClose={handleDialogClose} />
      <StyledAppBar position="static">
        <Toolbar>
          <StyledIconButton
            edge="start"
            color="inherit"
            aria-label="open drawer"
            onClick={toggleDrawer(true)}
          >
            <MenuIcon />
          </StyledIconButton>
          <Drawer
            anchor="left"
            className="headerDrawer"
            open={openDrawer}
            onClose={toggleDrawer(false)}
          >
            <StyledList
              role="presentation"
              onClick={toggleDrawer(false)}
              onKeyDown={toggleDrawer(false)}
            >
              <ListItem button onClick={handleGoBack}>
                <ListItemIcon>
                  <HomeIcon />
                </ListItemIcon>
                <ListItemText primary="Home" />
              </ListItem>
            </StyledList>
          </Drawer>

          <StyledSearchDiv>
            <SearchIconDiv>
              <SearchIcon />
            </SearchIconDiv>

            <StyledForm aria-label="search form" onSubmit={handleSearchOnSubmit}>
              <StyledInputBase
                placeholder="Searchâ€¦"
                inputComponent={StyledInput}
                inputProps={{ 'aria-label': 'search' }}
                value={searchTerm}
                onChange={handleSearchOnChange}
              />
            </StyledForm>
          </StyledSearchDiv>

          <StyledGrowDiv />

          <StyledDarkModeDiv>
            <FormControlLabel
              label="Dark mode"
              control={
                <StyledSwitch
                  checked={darkMode}
                  onChange={switchDarkMode}
                  aria-label="dark mode switch"
                />
              }
            />
            <IconButton
              edge="end"
              data-testid="iconbutton-testid"
              aria-label="account of current user"
              aria-controls="menuId"
              aria-haspopup="true"
              onClick={handleMenuOpen}
              color="inherit"
            >
              <AccountCircleIcon />
            </IconButton>
            <StyledMenu
              id="menuId"
              data-testid="menu-testid"
              anchorEl={anchorEl}
              open={isMenuOpen}
              onClose={handleMenuClose}
              anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
              transformOrigin={{ vertical: 'top', horizontal: 'right' }}
              keepMounted
            >
              {false ? (
                <MenuItem onClick={handleMenuClose}>Log Out</MenuItem>
              ) : (
                <MenuItem onClick={handleDialogOpen}>Log In</MenuItem>
              )}
            </StyledMenu>
          </StyledDarkModeDiv>
        </Toolbar>
      </StyledAppBar>
    </>
  );
};

export default Header;
