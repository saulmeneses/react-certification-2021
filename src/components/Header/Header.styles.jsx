import styled, { css } from 'styled-components';
import { AppBar, IconButton, InputBase, Switch, Menu } from '@material-ui/core';

export const StyledGrowDiv = styled.div`
  flex-grow: 1;
`;

export const StyledAppBar = styled(AppBar)`
  && {
    background-color: ${({ theme }) => theme.appBar.backgroundColor};
  }
`;

export const StyledIconButton = styled(IconButton)`
  margin-right: 16px;
`;

export const StyledSearchDiv = styled.div`
  width: 100%;
  position: relative;
  border-radius: 4px;
  margin-right: 16px;
  margin-left: 0;
  background-color: #ffffff15;

  &:hover {
    background-color: #ffffff25;
  }

  @media (min-width: 600px) {
    margin-left: 24px;
    width: auto;
  }
`;

export const SearchIconDiv = styled.div`
  position: absolute;
  height: 100%;
  padding: 0px 16px;
  pointer-events: none;

  display: flex;
  justify-content: center;
  align-items: center;
`;

export const StyledForm = styled.form`
  width: 100%;
`;

export const StyledInputBase = styled(InputBase)`
  color: inherit;
`;

export const StyledInput = styled.input`
  && {
    width: 100%;
    padding: 8px 8px 8px 0px;
    padding-left: calc(1em + 32px);
    color: #ffff;

    @media (min-width: 960px) {
      width: 20ch;
    }
  }
`;

export const StyledDarkModeDiv = styled.div`
  display: none;

  @media (min-width: 960px) {
    display: flex;
  }
`;

export const StyledSwitch = styled(Switch)`
  ${(props) =>
    props.checked &&
    css`
      && {
        .Mui-checked {
          color: #19857b;
        }
        .Mui-checked + .MuiSwitch-track {
          background-color: #19857b;
        }
      }
    `};
`;

export const StyledMenu = styled(Menu)`
  .MuiPaper-root {
    color: ${({ theme }) => theme.body.color};
    background-color: ${({ theme }) => theme.body.backgroundColor};
  }
`;
