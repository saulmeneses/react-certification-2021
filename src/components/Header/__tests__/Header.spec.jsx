import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Header from '..';
import CustomProvider from '../../../providers/Custom';
import { StyledSwitch } from '../Header.styles';

const WrappedHeader = () => (
  <CustomProvider>
    <Header />
  </CustomProvider>
);

describe('<Header />', () => {
  it('renders search bar', () => {
    render(<WrappedHeader />);

    expect(screen.getByLabelText('search')).toBeInTheDocument();
  });

  it('typing on search bar', () => {
    render(<WrappedHeader />);

    const searchElement = screen.getByLabelText('search');
    const value = 'wizeline';

    fireEvent.change(searchElement, { target: { value } });
    expect(searchElement.value).toBe(value);
  });

  it('submits text in search bar', () => {
    const mockedSubmit = jest.fn(() => {});

    render(<WrappedHeader />);

    const formElement = screen.getByLabelText('search form');
    const inputElement = formElement.querySelector('input');
    const value = 'dog';

    formElement.onsubmit = mockedSubmit;

    fireEvent.change(inputElement, { target: { value } });
    fireEvent.submit(formElement);

    expect(inputElement.value).toBe(value);
    expect(mockedSubmit).toHaveBeenCalled();
  });

  it('renders dark mode switch', () => {
    render(<WrappedHeader />);

    expect(screen.getByLabelText('dark mode switch')).toBeInTheDocument();
  });

  it('executes dark mode switch onChange event', () => {
    let checked = false;
    const handleOnChange = jest.fn(() => {
      checked = !checked;
    });
    const ariaLabel = 'mock switch';

    render(
      <StyledSwitch aria-label={ariaLabel} checked={checked} onChange={handleOnChange} />
    );

    const element = screen
      .getByLabelText(ariaLabel)
      .querySelector('input[type="checkbox"]');

    fireEvent.click(element);
    expect(handleOnChange).toHaveBeenCalled();
    expect(checked).toBeTruthy();
  });

  it('renders and match checked switch', () => {
    const tree = renderer.create(<StyledSwitch checked />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('shows account menu', () => {
    const renderRes = render(<WrappedHeader />);

    const element = renderRes.container.querySelector('button[aria-controls="menuId"]');

    fireEvent.click(element);

    expect(element.querySelector('svg')).toBeInTheDocument();
    expect(screen.getByTestId('menu-testid')).toBeVisible();
  });
});
