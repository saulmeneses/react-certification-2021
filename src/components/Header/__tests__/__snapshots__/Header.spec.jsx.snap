// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders and match checked switch 1`] = `
<span
  className="MuiSwitch-root sc-pFZIQ FFIYX"
>
  <span
    aria-disabled={false}
    className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-21 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-22 Mui-checked"
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={null}
  >
    <span
      className="MuiIconButton-label"
    >
      <input
        checked={true}
        className="PrivateSwitchBase-input-24 MuiSwitch-input"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="MuiSwitch-thumb"
      />
    </span>
  </span>
  <span
    className="MuiSwitch-track"
  />
</span>
`;
